semtypes/src/test/java/io/ballerina/semtype/SemTypeCoreTest.java

```diff
@@ -299,4 +300,47 @@ public void testIntSubtypeWidenUnsigned() {
         Assert.assertEquals(intType2.ranges[0].min, 0L);
         Assert.assertEquals(intType2.ranges[0].max, 65535L);
     }
+
+    public SemType recursiveTuple(Env env, LambdaHolder f) {
+        ListDefinition def = new ListDefinition();
+        SemType t = def.getSemType(env);
+        List<SemType> members = f.op(env, t);
+        return def.define(env, members, PredefinedType.NEVER);
+    }
+
+    @Test
+    public void recTest() {
+        Env env = new Env();
+        LambdaHolder f1 = (e, t) -> Arrays.asList(PredefinedType.INT, Core.union(t, PredefinedType.NIL));
+        LambdaHolder f2 = (e, t) -> Arrays.asList(Core.union(PredefinedType.INT, PredefinedType.STRING),
+                Core.union(t, PredefinedType.NIL));
+        SemType t1 = recursiveTuple(env, f1);
```

Comment:
Let's inline the lambdas here, so that we don't need this `LambdaHolder` interface.
```java
SemType t1 = recursiveTuple(env, (e, t) -> Arrays.asList(PredefinedType.INT, Core.union(t, PredefinedType.NIL)));

```